driven/storage/adapter.go

```diff
@@ -513,45 +524,84 @@ func (sa *Adapter) updateGroup(clientID string, current *model.User, group *mode
-		primitive.E{Key: "research_open", Value: group.ResearchOpen},
-		primitive.E{Key: "research_consent_statement", Value: group.ResearchConsentStatement},
-		primitive.E{Key: "research_consent_details", Value: group.ResearchConsentDetails},
-		primitive.E{Key: "research_description", Value: group.ResearchDescription},
-		primitive.E{Key: "research_profile", Value: group.ResearchProfile},
-		primitive.E{Key: "attributes", Value: group.Attributes},
-	}
-	if group.Settings != nil {
-		setOperation = append(setOperation, primitive.E{Key: "settings", Value: group.Settings})
-	}
-
-	updateOperation := bson.D{
-		primitive.E{Key: "$set", Value: setOperation},
-	}
 
 	// transaction
 	err := sa.PerformTransaction(func(context TransactionContext) error {
+		if group.ResearchGroup && group.CanJoinAutomatically {
+			group.CanJoinAutomatically = false
+		}
+
+		setOperation := bson.D{
+			primitive.E{Key: "title", Value: group.Title},
+			primitive.E{Key: "privacy", Value: group.Privacy},
+			primitive.E{Key: "hidden_for_search", Value: group.HiddenForSearch},
+			primitive.E{Key: "description", Value: group.Description},
+			primitive.E{Key: "image_url", Value: group.ImageURL},
+			primitive.E{Key: "web_url", Value: group.WebURL},
+			primitive.E{Key: "membership_questions", Value: group.MembershipQuestions},
+			primitive.E{Key: "date_updated", Value: time.Now()},
+			primitive.E{Key: "authman_enabled", Value: group.AuthmanEnabled},
+			primitive.E{Key: "authman_group", Value: group.AuthmanGroup},
+			primitive.E{Key: "only_admins_can_create_polls", Value: group.OnlyAdminsCanCreatePolls},
+			primitive.E{Key: "can_join_automatically", Value: group.CanJoinAutomatically},
+			primitive.E{Key: "block_new_membership_requests", Value: group.BlockNewMembershipRequests},
+			primitive.E{Key: "attendance_group", Value: group.AttendanceGroup},
+			primitive.E{Key: "research_group", Value: group.ResearchGroup},
+			primitive.E{Key: "research_open", Value: group.ResearchOpen},
+			primitive.E{Key: "research_consent_statement", Value: group.ResearchConsentStatement},
+			primitive.E{Key: "research_consent_details", Value: group.ResearchConsentDetails},
+			primitive.E{Key: "research_description", Value: group.ResearchDescription},
+			primitive.E{Key: "research_profile", Value: group.ResearchProfile},
+		}
+		if group.Settings != nil {
+			setOperation = append(setOperation, primitive.E{Key: "settings", Value: group.Settings})
+		}
+
+		//
+		// Handle category and tags backward compatibility and legacy clients 355355
+		//
+		if group.Attributes != nil {
+			setOperation = append(setOperation, primitive.E{Key: "attributes", Value: group.Attributes})
+
+			category := group.GetNewCategory()
+			if category != nil {
+				setOperation = append(setOperation, primitive.E{Key: "category", Value: *category})
+			}
+
+			tags := group.GetNewTags()
+			if tags != nil {
+				setOperation = append(setOperation, primitive.E{Key: "tags", Value: tags})
+			}
```

Comment:
Shouldn't we try to get rid of the root level `category` and `tags` fields and migrate this data into the attributes so we don't need to maintain these duplicated copies in the database? 

To handle backwards compatibility we can populate the root level `category` and `tags` fields with the data from the respective fields in `attributes` before sending the object back to the client. This way when we have forced upgrade to the latest app it will be easy to just remove these fields and related logic rather than needing to handle the migration at that point.

Reply:
Hi @shurwit, I started in this direction initially, but we have existing clients who don't know the attribute stuff and new ones that don't care for the root elements. In other words some agreement must be taken at least temporarly - otherwise a forced client upgrade must be applied, which I think is a bad experience.